### PR TITLE
Add hostNAT to vrouter

### PR DESCRIPTION
**Pull request by jojimt**
Signed-off-by: Joji Mekkatt <joji@cisco.com>

Add hostNAT port to vrouter in order to enable an exit path from overlay.

### CODE REVIEW

**Comment by jojimt** [comment]
@DivyaVavili @rchirakk PTAL

File: ofnetAgent.go [hunk 1]
```diff
@@ -398,6 +410,18 @@ func (self *OfnetAgent) PacketRcvd(sw *ofctrl.OFSwitch, pkt *ofctrl.PacketIn) {
 	self.incrStats("PktRcvd")
 }
 
+// AddRoute adds a route into datapath
+func (self *OfnetAgent) AddRoute(vrfName, cidr, nhIP, nhMAC string, portNo uint32) error {
+	//return self.datapath.AddRoute(vrfName, cidr, nhIP, nhMAC, portNo)
```
Comment (rchirakk) [inline ofnetAgent.go]: please remove it if it is not used

File: vrouter.go [hunk 2]
```diff
@@ -397,6 +480,78 @@ func (self *Vrouter) UpdateLocalEndpoint(endpoint *OfnetEndpoint, epInfo Endpoin
...
+	denyFlow.Next(self.ofSwitch.DropAction())
+
+	self.hostNATFlowDB.Set("inNATFlow", inNATFlow)
+	self.hostNATFlowDB.Set("inNATMiss", inNATMiss)
+	self.hostNATFlowDB.Set("denyFlow", denyFlow)
+	return nil
+}
+
+// RemoveHostPort sets up host access.
+func (self *Vrouter) RemoveHostPort(hp uint32) error {
```
Comment (DivyaVavili) [inline vrouter.go]: Check for nil hostNATInfo

File: vrouter.go [hunk 3]
```diff
@@ -397,6 +480,78 @@ func (self *Vrouter) UpdateLocalEndpoint(endpoint *OfnetEndpoint, epInfo Endpoin
...
+	inNATFlow, _ := self.inputTable.NewFlow(ofctrl.FlowMatch{
+		InputPort: self.hostNATInfo.PortNo,
+		Priority:  FLOW_MATCH_PRIORITY,
+	})
+	inNATFlow.Next(self.hostDNATTable)
+
+	// Packets from hostport that miss in IP table are dropped
+	inNATMiss, _ := self.ipTable.NewFlow(ofctrl.FlowMatch{
+		InputPort: self.hostNATInfo.PortNo,
+		Priority:  FLOW_FLOOD_PRIORITY,
```
Comment (DivyaVavili) [inline vrouter.go]: Should this be FLOW_MISS_PRIORITY?

File: vrouter.go [hunk 4]
```diff
@@ -1062,25 +1234,31 @@ func (self *Vrouter) processArp(pkt protocol.Ethernet, inPort uint32) {
...
-			endpoint := self.agent.getEndpointByID(endpointId)
-			if endpoint == nil {
-				// Look for a service entry for the target IP
-				proxyMac := self.svcProxy.GetSvcProxyMAC(arpHdr.IPDst)
-				if proxyMac == "" {
-					// If we dont know the IP address, dont send an ARP response
-					log.Debugf("Received ARP request for unknown IP: %v", arpHdr.IPDst)
-					self.agent.incrStats("ArpReqUnknownDest")
+			var tgtMac net.HardwareAddr
+			if inPort != 0 && inPort == self.hostNATInfo.PortNo {
```
Comment (DivyaVavili) [inline vrouter.go]: Include a check for nil self.hostNATInfo

File: vrouter.go [hunk 5]
```diff
@@ -397,6 +480,78 @@ func (self *Vrouter) UpdateLocalEndpoint(endpoint *OfnetEndpoint, epInfo Endpoin
...
+	if err != nil {
+		log.Errorf("Bad host route - %v", err)
+		return err
+	}
+
+	netMask := pflag.ParseIPv4Mask(daMask.String())
+	maskedIP := ipDa.Mask(netMask)
+
+	// Save the info
+	self.hostNATInfo = hp
```
Comment (DivyaVavili) [inline vrouter.go]: If this is intended to be stored in ofnet, it should be a pointer. Please update the function signatures also accordingly.

File: vrouter.go [hunk 6]
```diff
@@ -397,6 +480,78 @@ func (self *Vrouter) UpdateLocalEndpoint(endpoint *OfnetEndpoint, epInfo Endpoin
...
+			inNATMiss.(*ofctrl.Flow).Delete()
+		}
+		if denyFlow != nil {
+			denyFlow.(*ofctrl.Flow).Delete()
+		}
+
+		self.hostNATFlowDB.Remove("inNATFlow")
+		self.hostNATFlowDB.Remove("inNATMiss")
+		self.hostNATFlowDB.Remove("denyFlow")
+	}
```
Comment (DivyaVavili) [inline vrouter.go]: Set self.hostNATInfo to nil

File: vrouter.go [hunk 7]
```diff
@@ -61,7 +65,10 @@ type Vrouter struct {
 	portDnsFlowDb  cmap.ConcurrentMap        // Database of flow entries
 	vlanDb         map[uint16]*Vlan          // Database of known vlans
 
-	myRouterMac net.HardwareAddr // Router Mac to be used
+	myRouterMac   net.HardwareAddr   // Router Mac to be used
+	staticRoutes  cmap.ConcurrentMap // Statically inserted routes
```
Comment (DivyaVavili) [inline vrouter.go]: Don't see this getting used anywhere. Remove if not used.

File: vrouter.go [hunk 8]
```diff
@@ -300,11 +309,85 @@ func (self *Vrouter) AddLocalEndpoint(endpoint OfnetEndpoint) error {
...
+	dnatFlowKey := endpoint.HostPvtIP.String() + ".dnat"
+	hdFlow, _ := self.hostNATFlowDB.Get(dnatFlowKey)
+	if hdFlow != nil {
+		hdFlow.(*ofctrl.Flow).Delete()
+	}
+	snatFlowKey := endpoint.HostPvtIP.String() + ".snat"
+	hsFlow, _ := self.hostNATFlowDB.Get(snatFlowKey)
+	if hsFlow != nil {
+		hsFlow.(*ofctrl.Flow).Delete()
+	}
```
Comment (DivyaVavili) [inline vrouter.go]: Remove flows from hostNATFlowDB and as well

Comment (DivyaVavili) [inline vrouter.go]: I see that that the many of the flows in this function are not getting removed from their DB. Could you take care of those as well?

File: vrouter.go (hunk 7 above)
Comment (jojimt) [inline vrouter.go]: Good catch. Will removed.

File: vrouter.go (hunk 8 above)
Comment (jojimt) [inline vrouter.go]: @DivyaVavili Please see the three self.hostNATFlowDB.Remove at the end of the block.

File: vrouter.go (hunk 5 above)
Comment (jojimt) [inline vrouter.go]: @DivyaVavili why do you say it needs to be a pointer?

File: vrouter.go (hunk 3 above)
Comment (jojimt) [inline vrouter.go]: No, we need this to be higher than MISS because this is a specific rule applicable only to NAT port.

File: vrouter.go (hunk 6 above)
Comment (jojimt) [inline vrouter.go]: N/A because I am not a maintaining a pointer.

File: vrouter.go (hunk 4 above)
Comment (jojimt) [inline vrouter.go]: See above.

File: vrouter.go (hunk 2 above)
Comment (jojimt) [inline vrouter.go]: hostNATInfo is not a pointer -- hence can't be nil.

File: vrouter.go (hunk 5 above)
Comment (DivyaVavili) [inline vrouter.go]: I thought I saw portNo in hostNATInfo getting manipulated here. Double checked again and saw that it was only during the remove. If we are not modifying anything here and not maintaining its reference in netplugin, a pointer is not needed. Please ignore my nil check comments as well in that case.

File: vrouter.go (hunk 8 above)
Comment (DivyaVavili) [inline vrouter.go]: I see the Remove calls in RemoveHostPort function only. Am I overlooking anything here?

Comment (jojimt) [inline vrouter.go]: You are not, I was... fixing it now.

**Comment by jojimt** [comment]
@DivyaVavili addressed your comments. PTAL.

File: vrouter.go [hunk 9]
```diff
@@ -300,11 +307,87 @@ func (self *Vrouter) AddLocalEndpoint(endpoint OfnetEndpoint) error {
 		}
 	}
 
+	if endpoint.HostPvtIP.String() != "<nil>" && self.hostNATInfo.PortNo != 0 {
```
Comment (rchirakk) [inline vrouter.go]: HostPvtIp.IsUnspecified() may be a better option

File: vrouter.go (hunk 5 above)
Comment (rchirakk) [inline vrouter.go]: is there any option supported other than Kind == "NAT" ?

File: vrouter.go (hunk 9 above)
Comment (jojimt) [inline vrouter.go]: Based on the documentation, it does not seem to be the same thing. It it's uninitialized, String will return <nil>.

File: ofnet_route_test.go [hunk 10]
```diff
@@ -16,9 +16,22 @@ import (
 	"google.golang.org/grpc"
 )
 
+func verifyHostNAT(t *testing.T, flowList []string, tableId int, flow string, expectFlow bool) {
+	if expectFlow != ofctlFlowMatch(flowList, tableId, flow) {
+		if expectFlow {
+			t.Errorf("Expected %s in table %d -- not found", flow, tableId)
+		} else {
+			t.Errorf("Unxpected %s in table %d", flow, tableId)
```
Comment (DivyaVavili) [inline ofnet_route_test.go]: very minor nit: spell check

File: vrouter.go (hunk 5 above)
Comment (jojimt) [inline vrouter.go]: As of now, no. But, it could be leveraged in the future to add a different kind of host port.

**Pull request merged**